File: backend/src/lib/crypto/hmac/hmac.ts
import { crypto } from "@app/lib/crypto/cryptography";
import { BadRequestError } from "@app/lib/errors";

import { HmacAlgorithm, THmacFns } from "./types";

const HMAC_ALGORITHM_CONFIG: Record<HmacAlgorithm, { hash: string; keyByteLength: number }> = {
  [HmacAlgorithm.HMAC_SHA_1]: { hash: "sha1", keyByteLength: 20 },
  [HmacAlgorithm.HMAC_SHA_224]: { hash: "sha224", keyByteLength: 28 },
  [HmacAlgorithm.HMAC_SHA_256]: { hash: "sha256", keyByteLength: 32 },
  [HmacAlgorithm.HMAC_SHA_384]: { hash: "sha384", keyByteLength: 48 },
  [HmacAlgorithm.HMAC_SHA_512]: { hash: "sha512", keyByteLength: 64 }
};

export const getHmacAlgorithmConfig = (algorithm: HmacAlgorithm) => {
  const config = HMAC_ALGORITHM_CONFIG[algorithm];
  if (!config) {
    throw new BadRequestError({ message: `Unsupported HMAC algorithm: ${algorithm as string}` });
  }
  return config;
};

/**
 * Service for HMAC (RFC 2104) message authentication code generation and verification.
 *
 * @param algorithm The HMAC algorithm, which also determines the hash function and generated key length.
 * @returns Object with generateMac, verifyMac, generateKeyMaterial and getKeyByteLength functions.
 */
export const hmacService = (algorithm: HmacAlgorithm): THmacFns => {
  const { hash, keyByteLength } = getHmacAlgorithmConfig(algorithm);

  const generateMac = (data: Buffer, key: Buffer) => crypto.nativeCrypto.createHmac(hash, key).update(data).digest();

  const verifyMac = (data: Buffer, mac: Buffer, key: Buffer) => {
    const expected = crypto.nativeCrypto.createHmac(hash, key).update(data).digest();
    if (expected.length !== mac.length) {
      return false;
    }
    return crypto.nativeCrypto.timingSafeEqual(expected, mac);
  };

  const generateKeyMaterial = () => crypto.randomBytes(keyByteLength);

  const getKeyByteLength = () => keyByteLength;

  return { generateMac, verifyMac, generateKeyMaterial, getKeyByteLength };
};

File: backend-go/pkg/lib/hmac/hmac.go
package hmac

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha1"
	"crypto/sha256"
	"crypto/sha512"
	"fmt"
	"hash"
)

type HmacAlgorithm string

const (
	HMACSHA1   HmacAlgorithm = "HMAC_SHA_1"
	HMACSHA224 HmacAlgorithm = "HMAC_SHA_224"
	HMACSHA256 HmacAlgorithm = "HMAC_SHA_256"
	HMACSHA384 HmacAlgorithm = "HMAC_SHA_384"
	HMACSHA512 HmacAlgorithm = "HMAC_SHA_512"
)

type hmacConfig struct {
	keyByteLength uint8
	hash          string
}

var HMAC_ALGORITHM_CONFIG = map[HmacAlgorithm]hmacConfig{
	HMACSHA1: {
		hash:          "sha1",
		keyByteLength: 20,
	},
	HMACSHA224: {
		hash:          "sha224",
		keyByteLength: 28,
	},
	HMACSHA256: {
		hash:          "sha256",
		keyByteLength: 32,
	},
	HMACSHA384: {
		hash:          "sha384",
		keyByteLength: 48,
	},
	HMACSHA512: {
		hash:          "sha512",
		keyByteLength: 64,
	},
}

func getHash(hashName string) (func() hash.Hash, error) {
	switch hashName {
	case "sha1":
		return sha1.New, nil
	case "sha224":
		return sha256.New224, nil
	case "sha256":
		return sha256.New, nil
	case "sha384":
		return sha512.New384, nil
	case "sha512":
		return sha512.New, nil
	default:
		return nil, fmt.Errorf("unsupported hash: %s", hashName)
	}
}

func GetHmacAlgorithmConfig(algorithm HmacAlgorithm) (hmacConfig, bool) {
	config, ok := HMAC_ALGORITHM_CONFIG[algorithm]
	return config, ok
}

func GenerateMAC(algorithm HmacAlgorithm, data, key []byte) ([]byte, error) {
	config, ok := HMAC_ALGORITHM_CONFIG[algorithm]
	if !ok {
		return nil, fmt.Errorf("unsupported HMAC algorithm: %s", algorithm)
	}

	hashFn, err := getHash(config.hash)
	if err != nil {
		return nil, err
	}

	mac := hmac.New(hashFn, key)
	_, _ = mac.Write(data)

	return mac.Sum(nil), nil
}

func VerifyMAC(algorithm HmacAlgorithm, data, expectedMAC, key []byte) (bool, error) {
	actualMAC, err := GenerateMAC(algorithm, data, key)
	if err != nil {
		return false, err
	}

	return hmac.Equal(actualMAC, expectedMAC), nil
}

func GenerateKeyMaterial(algorithm HmacAlgorithm) ([]byte, error) {
	config, ok := HMAC_ALGORITHM_CONFIG[algorithm]
	if !ok {
		return nil, fmt.Errorf("unsupported HMAC algorithm: %s", algorithm)
	}

	key := make([]byte, config.keyByteLength)

	if _, err := rand.Read(key); err != nil {
		return nil, err
	}

	return key, nil
}

func GetKeyByteLength(algorithm HmacAlgorithm) (uint8, error) {
	config, ok := HMAC_ALGORITHM_CONFIG[algorithm]
	if !ok {
		return 0, fmt.Errorf("unsupported HMAC algorithm: %s", algorithm)
	}

	return config.keyByteLength, nil
}
